### DISCUSSION TITLE
quarto markdown list and blank line

### DISCUSSION THREAD

**Discussion by eijoac** [quarto-dev/quarto-cli#4650] [Q&A]
I wonder whether the default blank line (or space) between list and list sub-items can be made consistent in Quarto markdown rendering. (See a similar discussion in the R community site too, https://community.rstudio.com/t/quarto-markdown-list-and-blank-line/160693/7)

Example 1.

```
---
title: "test"
format: html
editor: visual
---

### Part 5

* unordered list
* unordered list
    + sub-item 1
    + sub-item 2
        - sub-sub-item 1
    + sub-time 3
* unordered list
* unordered list
``` 
It renders to,

![image](https://user-images.githubusercontent.com/10935026/223212981-6f811f74-7289-476b-bbf6-45d95df5a5fe.png)

Example 2

```
---
title: "test"
format: html
editor: visual
---

### Part 5

* unordered list

* unordered list
    + sub-item 1
    + sub-item 2
        - sub-sub-item 1
    + sub-time 3
* unordered list
* unordered list
```

It renders to,

![image](https://user-images.githubusercontent.com/10935026/223213499-78dc44f0-0ff3-4007-ab6c-07316815a6ed.png)

**Comment by cscheid** [comment]
Thanks for the feedback! There's definitely some weird things happening with the spacing, but unfortunately the story is a bit more complicated. Let me explain. There are two causes of spacing here:

## "Plain" bullet list items vs "Paragraph" bullet list items

In Pandoc, the content of a bulleted list is interpreted in one of two things: "plain" content that should have no spacing between items, and "paragraph" content that _should_ have spacing between items.

Crucially, Pandoc supports _only one of these_ at any one bulleted list. 

This gets admittedly very unintuitive because for the most part, Markdown doesn't care about empty line. But in this case, _any one_ empty line between bullet list items means that Pandoc thinks you want spacing among _all items of that list_. So in your second example, I think you meant to ask Pandoc "please add a space here, but nowhere else in this list". Unfortunately, that's not possible, and so Pandoc adds spacing between all of the items in your list.

## Margins at the bottom of bullet lists

We currently add `1rem` of margin at the bottom of any bullet list:

<img width="288" alt="image" src="https://user-images.githubusercontent.com/285675/223216761-6316dd3a-e5af-4c03-8e9c-534708ea0379.png">

That happens to be the same amount of space we add between paragraphs (and margins don't _add_, it's more like you get the largest of the applicable margins), so it's confusing when there exists "Paragraph" bullet lists together with the end of a bullet list.

## A (partial) solution

I think this latter problem is a bug in our .css, thanks for catching it! But the fix is for nested bullet lists to not have the bottom margin applied to them.

Unfortunately, I think you were hoping to get paragraph breaks wherever you have a line break, and nowhere else. Unfortunately, that's not possible with the way that Pandoc interprets markdown. You can still do it with custom CSS and spans, but it will require manual markup.

## The Pandoc details

```
$ cat bullet-list-styles.md
## Plain bullet lists

- Item 1
- Item 2
  - Item 3
  - Item 4
- Item 5

## Paragraph bullet lists

- Item 1

- Item 2
  - Item 3
  - Item 4
- Item 5
$ quarto pandoc -f markdown -t native bullet-list-styles.md
[ Header
    2
    ( "plain-bullet-lists" , [] , [] )
    [ Str "Plain" , Space , Str "bullet" , Space , Str "lists" ]
, BulletList
    [ [ Plain [ Str "Item" , Space , Str "1" ] ]
    , [ Plain [ Str "Item" , Space , Str "2" ]
      , BulletList
          [ [ Plain [ Str "Item" , Space , Str "3" ] ]
          , [ Plain [ Str "Item" , Space , Str "4" ] ]
          ]
      ]
    , [ Plain [ Str "Item" , Space , Str "5" ] ]
    ]
, Header
    2
    ( "paragraph-bullet-lists" , [] , [] )
    [ Str "Paragraph"
    , Space
    , Str "bullet"
    , Space
    , Str "lists"
    ]
, BulletList
    [ [ Para [ Str "Item" , Space , Str "1" ] ]
    , [ Para [ Str "Item" , Space , Str "2" ]
      , BulletList
          [ [ Plain [ Str "Item" , Space , Str "3" ] ]
          , [ Plain [ Str "Item" , Space , Str "4" ] ]
          ]
      ]
    , [ Para [ Str "Item" , Space , Str "5" ] ]
    ]
]
```

**Reply by cscheid** [reply]
We just fixed this on `main`, and now the nesting is at least symmetrical:

<img width="249" alt="image" src="https://user-images.githubusercontent.com/285675/223233312-eccb59ad-c213-41c3-b965-3942c575bc28.png">

**Reply by eijoac** [reply]
Thanks. Don't fully understand the Pandoc issue, but the fix looks good. I would probably use the "no spaces on outer or inner". I wish I could also have the "spaces on outer but not inner" with one extra blank space removed (see the picture below), but I guess the Pandoc issue you mentioned prevent that from happening (unless I add custom css).

![image](https://user-images.githubusercontent.com/10935026/223240401-49ecdd46-cfff-4dd3-85c3-13763eb15e49.png)

**Reply by cscheid** [reply]
Notice that you can choose which of these you want. This is the source .qmd file I used (with the quarto version from `main`):

````
---
title: "test"
format: html
editor: visual
---

### No spaces on outer or inner

Some stuff that comes before.

* unordered list
* unordered list
    + sub-item 1
    + sub-item 2
        - sub-sub-item 1
    + sub-time 3
* unordered list
* unordered list

Some stuff that comes after.

### Spaces on outer but not inner

Some stuff that comes before.

* unordered list

* unordered list
    + sub-item 1
    + sub-item 2
        - sub-sub-item 1
    + sub-time 3
* unordered list
* unordered list

Some stuff that comes after.

### Spaces on inner but not outer

Some stuff that comes before.

* unordered list
* unordered list
    + sub-item 1

    + sub-item 2
        - sub-sub-item 1
    + sub-time 3
* unordered list
* unordered list

Some stuff that comes after.


### Spaces on inner and outer

Some stuff that comes before.

* unordered list

* unordered list
    + sub-item 1

    + sub-item 2
        - sub-sub-item 1
    + sub-time 3
* unordered list
* unordered list

Some stuff that comes after.
````

**Reply by eijoac** [reply]
Thanks again. By the way, how do I achieve "spaces on outer but not inner" with that one extra blank space removed (as shown in my last picture) using custom css or scss. I couldn't figure it out after some googling and tests. I admit I don't know much about css/scss. Thanks.

**Reply by cscheid** [reply]
> but I guess the Pandoc issue you mentioned prevent that from happening (unless I add custom css).

You can add the CSS if you want; I checked with the quarto team and we happen to prefer the style where the spacing is applied between the li text and the inner lists. But if you prefer your behavior, use this:

````
<style>
li > p:has(+ul) {
    margin-bottom: 0;
}
</style>
````

**Reply by mcanouil** [reply]
@cscheid by the way, the spacing is now similar to what is done on GitHub.

---

### No spaces on outer or inner

Some stuff that comes before.

* unordered list
* unordered list
    + sub-item 1
    + sub-item 2
        - sub-sub-item 1
    + sub-time 3
* unordered list
* unordered list

Some stuff that comes after.

### Spaces on outer but not inner

Some stuff that comes before.

* unordered list

* unordered list
    + sub-item 1
    + sub-item 2
        - sub-sub-item 1
    + sub-time 3
* unordered list
* unordered list

Some stuff that comes after.

### Spaces on inner but not outer

Some stuff that comes before.

* unordered list
* unordered list
    + sub-item 1

    + sub-item 2
        - sub-sub-item 1
    + sub-time 3
* unordered list
* unordered list

Some stuff that comes after.


### Spaces on inner and outer

Some stuff that comes before.

* unordered list

* unordered list
    + sub-item 1

    + sub-item 2
        - sub-sub-item 1
    + sub-time 3
* unordered list
* unordered list

Some stuff that comes after.

**Reply by cscheid** [reply]
@mcanouil That's a good sanity check for us, thanks!
